src/stream.rs

```diff
@@ -0,0 +1,273 @@
+//! Provides a struct that implements logic to present a stream of disparate byte slices as a single stream of bytes
+
+use embedded_usb_pd::PdError;
+
+use crate::trace;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct ReadState {
+    /// Total number of bytes to read
+    pub total: usize,
+    /// Number of bytes already read for the current read operation
+    pub current: usize,
+}
+
+impl ReadState {
+    /// Create a new read state with the given total length
+    pub fn new(total: usize) -> Self {
+        ReadState { total, current: 0 }
+    }
+}
+
+/// Data stream state
+enum State {
+    /// Skip data until we reach the given byte
+    Seek(usize),
+    /// Consume the given number of bytes
+    Read(ReadState),
+}
+
+/// Data stream implementation
+pub struct Stream {
+    /// Curent offset within the overall stream
+    global_offset: usize,
+    /// Current operation state
+    state: State,
+}
+
+impl Stream {
+    /// Create a new stream
+    pub fn new(start_at: usize) -> Self {
+        Stream {
+            global_offset: 0,
+            state: State::Seek(start_at),
+        }
+    }
+
+    /// Returns if we're currently seeking in the stream
+    pub fn is_seeking(&self) -> bool {
+        matches!(self.state, State::Seek(_))
+    }
+
+    /// Returns if we're currently reading from the stream
+    pub fn is_reading(&self) -> bool {
+        matches!(self.state, State::Read(_))
+    }
+
+    /// Start a seek operation to the given byte
+    pub fn start_seek(&mut self, offset: usize) -> Result<(), PdError> {
+        if offset < self.global_offset {
+            return Err(PdError::InvalidParams);
+        }
+        self.state = State::Seek(offset);
+        Ok(())
+    }
+
+    /// Supply a byte slice to the stream
+    /// Returns (remaining data, bytes needed to reach the target)
+    pub fn seek_bytes<'a>(&mut self, data: &'a [u8]) -> Result<(&'a [u8], usize), PdError> {
+        let target = match self.state {
+            State::Seek(target) => target,
+            State::Read(_) => Err(PdError::InvalidMode)?,
+        };
+
+        if self.global_offset + data.len() < target {
+            trace!(
+                "Still waiting for target byte: {:#x} current: {:#x}",
+                target,
+                self.global_offset
+            );
+            // Still waiting for a particular byte
+            self.global_offset += data.len();
+            return Ok((&data[0..0], target - self.global_offset));
+        }
+
+        // Drop any bytes before the target
+        trace!("Seeked to {} {}", target, self.global_offset);
+        let skip = target - self.global_offset;
+        trace!(
+            "Data length: {} current: {} target: {}, skipping {} bytes",
+            data.len(),
+            self.global_offset,
+            target,
+            skip
+        );
+        self.global_offset = target;
+        Ok((&data[skip..], 0))
```

Comment:
I think all of this can be simplified with `slice::split_at_checked()`
```suggestion
        if let Some((_, remaining)) = data.split_at_checked(target - self.global_offset) {
            trace!(
                "Data length: {} current: {} target: {}, skipping {} bytes",
                data.len(),
                self.global_offset,
                target,
                skip
            );
            
            self.global_offset = target;
            Ok((remaining, 0))
        } else {
            trace!(
                "Still waiting for target byte: {:#x} current: {:#x}",
                target,
                self.global_offset
            );

            // Still waiting for a particular byte
            self.global_offset += data.len();
            Ok((data, target - self.global_offset)
        }
```

Reply:
Done.